src/tools/cephfs/JournalTool.cc

```diff
@@ -492,6 +539,243 @@ int JournalTool::journal_reset()
+  {
+    dirfrag_t const &frag = *lp;
+    EMetaBlob::dirlump const &lump = metablob.lump_map.find(frag)->second;
+    lump._decode_bits();
+    object_t frag_oid = InodeStore::get_object_name(frag.ino, frag.frag, "");
+
+    dout(4) << "inspecting lump " << frag_oid.name << dendl;
+
+    // Check for presence of dirfrag object
+    uint64_t psize;
+    time_t pmtime;
+    r = io.stat(frag_oid.name, &psize, &pmtime);
+    if (r == -ENOENT) {
+      dout(4) << ": Frag object " << frag_oid.name << " did not exist, will create" << dendl;
+    } else if (r != 0) {
+      derr << "Unexpected error stat'ing frag " << frag_oid.name
+           << ": " << cpp_strerror(r) << dendl;
+      return r;
+    } else {
+      dout(4) << "Frag object " << frag_oid.name << " exists, will modify" << dendl;
+    }
+
+    // Update fnode in omap header of dirfrag object
+    bool write_fnode = false;
+    bufferlist old_fnode_bl;
+    r = io.omap_get_header(frag_oid.name, &old_fnode_bl);
+    if (r == -ENOENT) {
+      // Creating dirfrag from scratch
+      dout(4) << "failed to read OMAP header from directory fragment "
+        << frag_oid.name << " " << cpp_strerror(r) << dendl;
+      write_fnode = true;
+      // Note: creating the dirfrag *without* a backtrace, relying on
+      // MDS to regenerate backtraces in FSCK
+    } else if (r == 0) {
+      // Conditionally update existing omap header
+      fnode_t old_fnode;
+      bufferlist::iterator old_fnode_iter = old_fnode_bl.begin();
+      old_fnode.decode(old_fnode_iter);  // FIXME handle decode error and overwrite
+      dout(4) << "frag " << frag_oid.name << " fnode old v" <<
+        old_fnode.version << " vs new v" << lump.fnode.version << dendl;
+      write_fnode = old_fnode.version < lump.fnode.version;
+    } else {
+      // Unexpected error
+      dout(4) << "failed to read OMAP header from directory fragment "
+        << frag_oid.name << " " << cpp_strerror(r) << dendl;
+      return r;
+    }
+
+    if (write_fnode && !dry_run) {
+      dout(4) << "writing fnode to omap header" << dendl;
+      bufferlist fnode_bl;
+      lump.fnode.encode(fnode_bl);
+      r = io.omap_set_header(frag_oid.name, fnode_bl);
+      if (r != 0) {
+        derr << "Failed to write fnode for frag object " << frag_oid.name << dendl;
+        return r;
+      }
+    }
+
+    // Try to get the existing dentry
+    list<ceph::shared_ptr<EMetaBlob::fullbit> > const &fb_list = lump.get_dfull();
+    for (list<ceph::shared_ptr<EMetaBlob::fullbit> >::const_iterator fbi = fb_list.begin(); fbi != fb_list.end(); ++fbi) {
+      EMetaBlob::fullbit const &fb = *(*fbi);
+
+      // Get a key like "foobar_head"
+      std::string key;
+      dentry_key_t dn_key(fb.dnlast, fb.dn.c_str());
+      dn_key.encode(key);
+
+      // See if the dentry is present
+      std::set<std::string> keys;
+      keys.insert(key);
+      std::map<std::string, bufferlist> vals;
+      r = io.omap_get_vals_by_keys(frag_oid.name, keys, &vals);
```

Comment:
Wait, this is a separate IO for each dentry in EMetablob? Can't we at least grab them all in a single IO? (I'm pretty sure you can do that just by inserting each one into the set of keys and the omap_get_vals_by_keys function will grab each of them for you individually?)